src/com/maddyhome/idea/vim/extension/textobjindent/VimIndentObject.java

```diff
@@ -0,0 +1,220 @@
+import static com.maddyhome.idea.vim.extension.VimExtensionFacade.putKeyMapping;
+import static com.maddyhome.idea.vim.group.visual.VisualGroupKt.vimSetSelection;
+import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;
+
+/**
+ * Port of vim-indent-object:
+ * https://github.com/michaeljsmith/vim-indent-object
+ *
+ * <p>
+ * vim-indent-object provides these text objects based on the cursor line's indentation:
+ * <ul>
+ *   <li><code>ai</code> <b>A</b>n <b>I</b>ndentation level and line above.</li>
+ *   <li><code>ii</code> <b>I</b>nner <b>I</b>ndentation level (no line above).</li>
+ *   <li><code>aI</code> <b>A</b>n <b>I</b>ndentation level and lines above and below.</li>
+ *   <li><code>iI</code> <b>I</b>nner <b>I</b>ndentation level (no lines above and below). Synonym of <code>ii</code></li>
+ * </ul>
+ *
+ * See also the reference manual for more details at:
+ * https://github.com/michaeljsmith/vim-indent-object/blob/master/doc/indent-object.txt
+ *
+ * @author Shrikant Kandula (@sharat87)
+ */
+public class VimIndentObject implements VimExtension {
+
+  @Override
+  public @NotNull
+  String getName() {
+    return "textobj-indent";
+  }
+
+  @Override
+  public void init() {
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-ai"), getOwner(),
+      new IndentObject(true, false), false);
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-aI"), getOwner(),
+      new IndentObject(true, true), false);
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-ii"), getOwner(),
+      new IndentObject(false, false), false);
+
+    putKeyMapping(MappingMode.XO, parseKeys("ai"), getOwner(), parseKeys("<Plug>textobj-indent-ai"), true);
+    putKeyMapping(MappingMode.XO, parseKeys("aI"), getOwner(), parseKeys("<Plug>textobj-indent-aI"), true);
+    putKeyMapping(MappingMode.XO, parseKeys("ii"), getOwner(), parseKeys("<Plug>textobj-indent-ii"), true);
+  }
+
+  static class IndentObject implements VimExtensionHandler {
+    final boolean includeAbove;
+    final boolean includeBelow;
+
+    IndentObject(boolean includeAbove, boolean includeBelow) {
+      this.includeAbove = includeAbove;
+      this.includeBelow = includeBelow;
+    }
+
+    static class IndentObjectHandler extends TextObjectActionHandler {
+      final boolean includeAbove;
+      final boolean includeBelow;
+
+      IndentObjectHandler(boolean includeAbove, boolean includeBelow) {
+        this.includeAbove = includeAbove;
+        this.includeBelow = includeBelow;
+      }
+
+      @Override
+      public @Nullable
+      TextRange getRange(@NotNull Editor editor, @NotNull Caret caret, @NotNull DataContext context,
+                         int count, int rawCount, @Nullable Argument argument) {
+        final int caretLineNum = caret.getCaretModel().getVisualPosition().getLine();
+        String content = editor.getDocument().getText();
```

Comment:
Is it possible to operate on `CharSequence` instead of the string? IJ uses a custom implementation of the `CharSequence` to store the code and converting it to the String may be a heavy operation.